### PR TITLE
added strict type guard for onRemove prop

### PR DESCRIPTION
**Pull request by GitJustinLau**
# Pull Request Template

## Description
onRemove is a prop which is only required and useful when addView = false. However, it can currently be input when addView=true. This PR changes the typing so that onRemove cannot be passed when addView=true

## Link to JIRA ticket
[Include the link here.](https://implentio.atlassian.net/jira/software/projects/IUT/boards/10?selectedIssue=IUT-35)

## Testing Instructions
Outline the steps required to test your changes effectively (Replace the following to what makes sense for your PR):
- Fetch and checkout my branch `git fetch && git checkout <fix/IUT-35/Update-FilterChip-w-type-guard>`
- Go to src/components/FilterChip/FilterChip.tsx
- on the bottom of the page, copy and paste: `<FilterChip addView={false} value="value" filter="filter: " onRemove={()=> console.log("test")} />`
- check if typescript throws type errors when you add and remove props and when you toggle addView to true/false

## Checklist:
Check what is applicable by putting an 'x' in between the []:
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have tested my code thoroughly.
- [ ] I made documentation changes (optional).

### CODE REVIEW

File: src/components/FilterChip/FilterChip.stories.tsx [hunk 1]
```diff
@@ -16,5 +16,6 @@ export const filterValue: Story = {
     filter: "Filter: ",
     value: "Value",
     addView: false,
+    onRemove: () => console.log("test")
```
Comment (launchupio) [inline src/components/FilterChip/FilterChip.stories.tsx:19]: Use actions for action related functions being called so we can leverage storybook's action tab.

At top of file:
`import { action } from '@storybook/addon-actions';`

```suggestion
    onRemove: () => action('onRemove')
```

more on it here: https://storybook.js.org/docs/essentials/actions

File: src/components/FilterChip/FilterChip.tsx [hunk 2]
```diff
@@ -14,37 +14,39 @@ const FilterChipVariants = cva(
...
   addView,
-  filter,
   value,
-  onRemove,
   ...props
 }: FilterChipProps) => {
   return (
     <div className={cn(FilterChipVariants({ className, addView }))} {...props}>
       {addView && value}
       {!addView && (
```
Comment (launchupio) [inline src/components/FilterChip/FilterChip.tsx:40]: I know it's part of the former PR of yours that already got okayed, but it looks like we missed this. We should simplify this to ternary to show it's outcome more clearly:
```
{ addView ? value : <p></p> }
~``

**Review by launchupio — CHANGES_REQUESTED** [review]
Please see comments! Make sure you start using actions please!

File: src/components/FilterChip/FilterChip.stories.tsx (hunk 1 above)
Comment (GitJustinLau) [reply src/components/FilterChip/FilterChip.stories.tsx:19]: change made

File: src/components/FilterChip/FilterChip.tsx (hunk 2 above)
Comment (GitJustinLau) [reply src/components/FilterChip/FilterChip.tsx]: ok change added

File: src/components/FilterChip/FilterChip.tsx [hunk 3]
```diff
@@ -14,39 +14,43 @@ const FilterChipVariants = cva(
...
+export type FilterChipProps = React.HTMLAttributes<HTMLDivElement> &
+  VariantProps<typeof FilterChipVariants> &
+  (addView | filterValue);
 
 const FilterChip = ({
   className,
   addView,
-  filter,
   value,
-  onRemove,
```
Comment (launchupio) [inline src/components/FilterChip/FilterChip.tsx:32]: taking this out causes this warning: 
Screenshot 2024-02-08 at 4 31 51 AM

If you keep this in, you'll need to edit FilterChipProps, as I've commented above.

File: src/components/FilterChip/FilterChip.tsx [hunk 4]
```diff
@@ -14,39 +14,43 @@ const FilterChipVariants = cva(
...
+  addView: false;
   filter: string;
   value: string;
-  addView: boolean;
   onRemove?: () => void;
-}
+};
+export type FilterChipProps = React.HTMLAttributes<HTMLDivElement> &
+  VariantProps<typeof FilterChipVariants> &
+  (addView | filterValue);
```
Comment (launchupio) [inline src/components/FilterChip/FilterChip.tsx:29]: If you leave in onRemove in line 32, you probably need to switch this around to something like:

```suggestion
export type FilterChipProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof FilterChipVariants> &
  (addView | filterValue)  & {
    onRemove?: () => void; 
  };
```

**Review by launchupio — CHANGES_REQUESTED** [review]

File: src/components/FilterChip/FilterChip.stories.tsx [hunk 5]
```diff
@@ -16,5 +15,6 @@ export const filterValue: Story = {
     filter: "Filter: ",
     value: "Value",
     addView: false,
+    onRemove: () => action('onRemove')
```
Comment (launchupio) [inline src/components/FilterChip/FilterChip.stories.tsx:18]: i think i originally asked for `onRemove: () => action('onRemove')` but it doesn't work. check docs. i think it should instead be `onRemove: action('onRemove')` instead perhaps

File: src/components/FilterChip/FilterChip.tsx (hunk 3 above)
Comment (launchupio) [reply src/components/FilterChip/FilterChip.tsx:32]: Never mind the idea of keeping it in. You can take it out like you had, but then have to deal with your current `...props` situation. I'll make comments elsewhere for an idea.

File: src/components/FilterChip/FilterChip.tsx [hunk 6]
```diff
@@ -14,39 +14,43 @@ const FilterChipVariants = cva(
...
 const FilterChip = ({
   className,
   addView,
-  filter,
   value,
-  onRemove,
   ...props
 }: FilterChipProps) => {
   return (
     <div className={cn(FilterChipVariants({ className, addView }))} {...props}>
```
Comment (launchupio) [inline src/components/FilterChip/FilterChip.tsx:38]: Dont' take verbatim, but something like this?

```suggestion
   let onRemoveProp;
  let restProps = props;
  
  if('onRemove' in props) {
    const { onRemove, ...otherProps } = props;
    onRemoveProp = onRemove;
    restProps = otherProps;
  }

  return (
    <div className={cn(FilterChipVariants({ className, addView }), className)} {...restProps}>
```

File: src/components/FilterChip/FilterChip.tsx [hunk 7]
```diff
@@ -14,39 +14,43 @@ const FilterChipVariants = cva(
...
+              <span className="font-regular">{value}</span>
+            </p>
+          ) &&
+          (props as filterValue).onRemove && (
+            <OnRemove
+              size="md"
+              className={`group-hover:text-icons-default`}
+              onClick={(props as filterValue).onRemove}
+            />
+          )}
```
Comment (launchupio) [inline src/components/FilterChip/FilterChip.tsx:53]:
```suggestion
{addView ? (
        value
      ) : (
        <>
          <p>
            {('filter' in props ? props.filter : '')} <span className="font-regular">{value}</span>
          </p>
          {'onRemove' in props && (
            <OnRemove
              size="md"
              className="group-hover:text-icons-default"
              onClick={onRemoveProp}
            />
          )}
        </>
      )}
```

**Review by launchupio — CHANGES_REQUESTED** [review]
Made some errors in my requested changes, and now i'm suggesting some other solutions as examples. You don't need to follow the ideas i wrong out though.

File: src/components/FilterChip/FilterChip.stories.tsx (hunk 5 above)
Comment (GitJustinLau) [reply src/components/FilterChip/FilterChip.stories.tsx:18]: Ok will confirm

Comment (GitJustinLau) [reply src/components/FilterChip/FilterChip.stories.tsx:18]: seems like action is for triggering an action when an event happens. The onRemove prop is for passing down a callback function. I think they have separate purposes, lmk.

image

File: src/index.css [hunk 8]
```diff
@@ -48,7 +48,7 @@
     --stroke-brand-muted: 249 58% 81%;
     --stroke-brand-disabled: 240 59% 15%;
 
-    --bg-brand-disabled: 249 58% 95%;     /* Icon Button */
```
Comment (launchupio) [inline src/index.css:51]: @GitJustinLau just want to confirm if deleting this is intentional?

Comment (GitJustinLau) [reply src/index.css:51]: this is the same variable i am using that has replaced the one removed: 
image

**Review by launchupio — APPROVED** [review]
👨 🚀 🌔 

Gj!

**Pull request merged**